### PR TITLE
enhance: enable balance channel in querycoord

### PR DESCRIPTION
**Pull request by weiliu1031**
issue: #23726

/kind improvement

1. enable auto balance channel between nodes in querycoord
2. make `genSegmentPlan` reuse the `AssignSegment` logic
3. make `genChannelPlan` reuse the `AssignChannel` logic

### CODE REVIEW

**Comment by weiliu1031** [comment]
rerun ut

File: internal/querycoordv2/balance/rowcount_based_balancer.go [hunk 1]
```diff
@@ -67,7 +69,35 @@ func (b *RowCountBasedBalancer) AssignSegment(collectionID int64, segments []*me
...
+	}
+	queue := newPriorityQueue()
+	for _, item := range nodeItems {
+		queue.push(item)
+	}
+
+	plans := make([]ChannelAssignPlan, 0, len(channels))
+	for _, c := range channels {
+		// pick the node with the least channel num and allocate to it.
+		ni := queue.pop().(*nodeItem)
```
Comment (jiaoew1991) [inline internal/querycoordv2/balance/rowcount_based_balancer.go:85]: can `ni` be nil?

File: internal/querycoordv2/balance/rowcount_based_balancer.go [hunk 2]
```diff
@@ -67,7 +69,35 @@ func (b *RowCountBasedBalancer) AssignSegment(collectionID int64, segments []*me
 	return plans
 }
 
-func (b *RowCountBasedBalancer) convertToNodeItems(nodeIDs []int64) []*nodeItem {
+func (b *RowCountBasedBalancer) AssignChannel(channels []*meta.DmChannel, nodes []int64) []ChannelAssignPlan {
```
Comment (jiaoew1991) [inline internal/querycoordv2/balance/rowcount_based_balancer.go:72]: this function need some comments to explain the policy

File: internal/querycoordv2/balance/rowcount_based_balancer.go [hunk 3]
```diff
@@ -83,127 +113,150 @@ func (b *RowCountBasedBalancer) convertToNodeItems(nodeIDs []int64) []*nodeItem
...
+			offlineNodes = append(offlineNodes, nid)
 		} else if outboundNodes.Contain(nid) {
 			// if node is stop or transfer to other rg
-			log.RatedInfo(10, "meet outbound node, try to move out all segment/channel",
-				zap.Int64("collectionID", replica.GetCollectionID()),
-				zap.Int64("replicaID", replica.GetCollectionID()),
-				zap.Int64("node", nid),
-			)
-			stoppingNodesSegments[nid] = segments
+			log.RatedInfo(10, "meet outbound node, try to move out all segment/channel", zap.Int64("node", nid))
```
Comment (jiaoew1991) [inline internal/querycoordv2/balance/rowcount_based_balancer.go:152]: add a RateGroup

File: internal/querycoordv2/balance/rowcount_based_balancer.go (hunk 1 above)
Comment (weiliu1031) [reply internal/querycoordv2/balance/rowcount_based_balancer.go:85]: no, the queue won't push nil element

**Review by yah01 — COMMENTED** [review]
/lgtm

**Pull request merged**